kpops/utils/dict_ops.py

```diff
@@ -96,3 +101,38 @@ def generate_substitution(
     return update_nested(
         existing_substitution or {}, flatten_mapping(input, prefix, separator)
     )
+
+
+_sentinel_dict = {}
+
+
+class ImprovedTemplate(Template):
+    """Introduces the dot as an allowed character in placeholders."""
+
+    idpattern = r"(?a:[_a-z][_.a-z0-9]*)"
+
+    @override
+    def safe_substitute(self, mapping=_sentinel_dict, /, **kws) -> str:
+        if mapping is _sentinel_dict:
+            mapping = kws
+        elif kws:
+            mapping = _ChainMap(kws, mapping)
+
+        # Helper function for .sub()
+        def convert(mo: re.Match):
+            named = mo.group("named") or mo.group("braced")
+            if named is not None:
+                try:
+                    if "." not in named:
+                        return str(mapping[named])
+                    return str(mapping[named.replace(".", "__")])
```

Comment:
doesn't the `prepare_substitution` function already do the dot→dunder conversion?

Reply:
Update: Unsubstituted placeholders are left the same!

Reply:
I tried removing `prepare_substitution` and confirmed this

> This is for the input while prepare_substitution is for the substitution mapping.